src/lib/linux-api/src/signal.rs

```diff
@@ -0,0 +1,548 @@
+#[repr(i32)]
+pub enum LinuxSignal {
+    SIGHUP = bindings::SIGHUP as i32,
+    SIGINT = bindings::SIGINT as i32,
+    SIGQUIT = bindings::SIGQUIT as i32,
+    SIGILL = bindings::SIGILL as i32,
+    SIGTRAP = bindings::SIGTRAP as i32,
+    SIGABRT = bindings::SIGABRT as i32,
+    SIGBUS = bindings::SIGBUS as i32,
+    SIGFPE = bindings::SIGFPE as i32,
+    SIGKILL = bindings::SIGKILL as i32,
+    SIGUSR1 = bindings::SIGUSR1 as i32,
+    SIGSEGV = bindings::SIGSEGV as i32,
+    SIGUSR2 = bindings::SIGUSR2 as i32,
+    SIGPIPE = bindings::SIGPIPE as i32,
+    SIGALRM = bindings::SIGALRM as i32,
+    SIGTERM = bindings::SIGTERM as i32,
+    SIGSTKFLT = bindings::SIGSTKFLT as i32,
+    SIGCHLD = bindings::SIGCHLD as i32,
+    SIGCONT = bindings::SIGCONT as i32,
+    SIGSTOP = bindings::SIGSTOP as i32,
+    SIGTSTP = bindings::SIGTSTP as i32,
+    SIGTTIN = bindings::SIGTTIN as i32,
+    SIGTTOU = bindings::SIGTTOU as i32,
+    SIGURG = bindings::SIGURG as i32,
+    SIGXCPU = bindings::SIGXCPU as i32,
+    SIGXFSZ = bindings::SIGXFSZ as i32,
+    SIGVTALRM = bindings::SIGVTALRM as i32,
+    SIGPROF = bindings::SIGPROF as i32,
+    SIGWINCH = bindings::SIGWINCH as i32,
+    SIGIO = bindings::SIGIO as i32,
+    SIGPWR = bindings::SIGPWR as i32,
+    SIGSYS = bindings::SIGSYS as i32,
+}
+
+impl LinuxSignal {
+    const fn const_alias(from: u32, to: Self) -> Self {
+        if to as i32 != from as i32 {
+            // Can't use a format string here since this function is `const`
+            panic!("Incorrect alias")
+        }
+        to
+    }
+    pub const SIGIOT: Self = Self::const_alias(bindings::SIGIOT, Self::SIGABRT);
+    pub const SIGPOLL: Self = Self::const_alias(bindings::SIGPOLL, Self::SIGIO);
+    pub const SIGUNUSED: Self = Self::const_alias(bindings::SIGUNUSED, Self::SIGSYS);
+}
+
+bitflags::bitflags! {
+    #[repr(transparent)]
+    #[derive(Copy, Clone, Debug, Default)]
+    pub struct LinuxSigActionFlags: core::ffi::c_int {
+        const NOCLDSTOP = bindings::SA_NOCLDSTOP as i32;
+        const NOCLDWAIT = bindings::SA_NOCLDWAIT as i32;
+        const NODEFER = bindings::SA_NODEFER as i32;
+        const ONSTACK = bindings::SA_ONSTACK as i32;
+        const RESETHAND = bindings::SA_RESETHAND as i32;
+        const RESTART = bindings::SA_RESTART as i32;
+        const RESTORER = bindings::SA_RESTORER as i32;
+        const SIGINFO = bindings::SA_SIGINFO as i32;
+    }
+}
+// We can't derive this since the bitflags field uses an internal type.
+unsafe impl VirtualAddressSpaceIndependent for LinuxSigActionFlags {}
+
+#[derive(Copy, Clone, VirtualAddressSpaceIndependent)]
+#[allow(non_camel_case_types)]
+#[repr(C)]
+pub struct linux_siginfo_t {
+    // We don't wrap bindings::siginfo_t directly,
+    // since this would introduce a dependency on the original system
+    // header file in the cbindgen'd version of this type, and requiring our C code
+    // to include the kernel headers results in naming conflicts.
```

Comment:
I don't understand this comment. Isn't the point of including the bindings for a specific kernel version that we don't need to worry about this? We just use the fields that exist in that specific kernel version?

Otherwise, I don't see why this wouldn't apply to all other structs in the Linux headers, and then we'd need to provide our own struct definitions for all Linux header structs.

Reply:
So this sounds like it will be an issue for each struct we define? But yeah I don't see a better alternative.

Reply:
One other option is to manually define the internals here, and validate that they match the bindings. e.g. that's the approach in `linux_sigaction`. It's a little more manual work up-front, but I suppose we shouldn't have to mess with it often.

```
pub struct linux_sigaction {
    // SAFETY: We do not dereference the pointers in this union, except from the
    // shim, where it is valid to do so.
    #[unsafe_assume_virtual_address_space_independent]
    u: LinuxSignalHandler,
    ksa_flags: LinuxSigActionFlags,
    // Rust guarantees that the outer Option doesn't change the size:
    // https://doc.rust-lang.org/std/option/index.html#representation
    //
    // SAFETY: We never dereference this field.
    #[unsafe_assume_virtual_address_space_independent]
    ksa_restorer: Option<extern "C" fn()>,
    ksa_mask: linux_sigset_t,
}

impl linux_sigaction {
    pub fn handler(&self) -> &LinuxSignalHandler {
        assert_eq!(
            memoffset::offset_of!(linux_sigaction, u),
            memoffset::offset_of!(bindings::sigaction, sa_handler)
        );
        &self.u
    }
}
```

Unfortunately I don't think there's a `const` equivalent of `memoffset::offset_of`, though maybe we could move those assertions to unit tests.

Reply:
I ended up doing a bit of a hybrid here. I added/kept explicit field definitions where feasible and added tests to validate field ordering. I left the `siginfo_t` mess of nested unions as `MaybeUninit<[u8;...]>` for the moment.

Reply:
Forgot to update this thread - we now use a script to mangle the names in the bindgen output so that we can directly wrap the bindgen types without causing naming conflicts in the cbindgen output.